### ISSUE TITLE
a bug in tensorflow/core/graph/gradient.cc

### ISSUE THREAD

**Issue by cuiguoxin** [tensorflow/tensorflow#1619]
in the above file, i think that in line 257 there is a error
for (const Edge* e : y->in_edges()) {
	if (e->IsControlEdge()) continue;
	BackpropAlongEdge(y_grad_node_outputs_[i], { e->src(), e->src_output() });
}
i think y->in_edges() should be y->out_edges(). please check this out as soon as possible, thank you

**Comment by cuiguoxin** [comment]
sorry,  i think just modify the above is not right . from line 254 to line 263 should be considered carefully to make it right.

**Comment by cuiguoxin** [comment]
if there is one node y has more than one output , then the node would be counted more than one time.

**Comment by zffchen78** [comment]
@cuiguoxin, you might have misunderstood the code. The code is initializing the backprop state starting from y(s). So it needs to look at y's inputs. If you think it's wrong,  do you mind constructing a test case to show it?

**Comment by cuiguoxin** [comment]
@zffchen78 ,does y_grad_node_outputs mean the gradient of y_node_outputs? To be clear, think the function y = 3 * x. Now i konw the node y, x and dy, and i want to compute the dx. The source code just put dy in x's backpropedGradient and i think it is wrong.

**Comment by andydavis1** [comment]
@cuiguoxin Thank you for your comments. I think that some of this confusion is due to this API being in an intermediate state (it currently just supports functions and not general graphs). As commented in the header file, this API is under construction, and will be changing frequently. Currently, the gradients passed in 'y_grad_node_outputs' are passed through to the input of the 'y_nodes', but these nodes are handled in a specialized way in the current implementation (i.e. function arg nodes). In a future version, the 'y_grad_node_outputs' will contain the gradient node outputs corresponding to the 'y_node_outputs', and will be propagated into the 'y_nodes' during backprop init.

**Comment by cuiguoxin** [comment]
@andydavis1 . Thanks. Does it mean that if I construct a graph and I want to auto differentiate the graph from the y to x, I still can't use the api in gradient.cc? Can you show me how to do it in c++ api?

**Comment by andydavis1** [comment]
I would not recommend using gradients.h/cc as it is currently under construction and will change frequently (as mentioned in the header file). If you need to auto differentiate your graph, for now, you need to use the python API (see code at tensorflow/python/ops/gradients.py)
